### ISSUE TITLE
cannot install kubelet on Raspbian using apt

### ISSUE THREAD

**Issue by SlyMarbo** [kubernetes/release#3707]
### What happened?

I'm trying to install `kubelet`, `kubeadm`, and `kubectl`, using apt on Raspbian, following the instructions in [the documentation](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/install-kubeadm/#installing-kubeadm-kubelet-and-kubectl).

I've successfully downloaded the signing key and APT list and I'm not getting any errors from `sudo apt update` or `sudo apt upgrade`. But trying `sudo apt install -y kubelet kubeadm kubectl` gives the following:

```
$ sudo apt install -y kubelet kubeadm kubectl
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
kubectl:arm64 is already the newest version (1.30.3-1.1).
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 kubelet:arm64 : Depends: socat:arm64 but it is not installable
                 Depends: ethtool:arm64 but it is not installable
                 Depends: conntrack:arm64 but it is not installable
                 Depends: libc6:arm64 but it is not going to be installed
E: Unable to correct problems, you have held broken packages.
```

I've manually installed `socat`, `ethtool`, `conntrack`, and `libc6` successfully, but the error continues. It looks like the cause is that `kubelet` depends explicitly on `socat:arm64` (which I can't install) and not the more generic `socat` (which is installed).

This is on a pretty fresh Raspbian install, so I don't believe I've done anything custom.

### What did you expect to happen?

APT to install the three packages successfully.

### How can we reproduce it (as minimally and precisely as possible)?

Install Raspbian, then follow the instructions in [the documentation](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/install-kubeadm/#installing-kubeadm-kubelet-and-kubectl):

```
$ sudo apt update
$ sudo apt install -y apt-transport-https ca-certificates curl gpg
$ curl -fsSL https://pkgs.k8s.io/core:/stable:/v1.30/deb/Release.key | sudo gpg --dearmor -o /etc/apt/keyrings/kubernetes-apt-keyring.gpg
$ echo 'deb [signed-by=/etc/apt/keyrings/kubernetes-apt-keyring.gpg] https://pkgs.k8s.io/core:/stable:/v1.30/deb/ /' | sudo tee /etc/apt/sources.list.d/kubernetes.list
$ sudo apt update
$ sudo apt install -y kubelet kubeadm kubectl
```

Note that I've used `apt` instead of `apt-get`, but there's no difference in effect.

### Kubernetes version

1.30 (but doesn't seem relevant)

### Cloud provider

N/A

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
PRETTY_NAME="Raspbian GNU/Linux 12 (bookworm)"
NAME="Raspbian GNU/Linux"
VERSION_ID="12"
VERSION="12 (bookworm)"
VERSION_CODENAME=bookworm
ID=raspbian
ID_LIKE=debian
HOME_URL="http://www.raspbian.org/"
SUPPORT_URL="http://www.raspbian.org/RaspbianForums"
BUG_REPORT_URL="http://www.raspbian.org/RaspbianBugs"
$ uname -a
Linux zeus 6.6.31+rpt-rpi-v8 kubernetes/kubernetes#1 SMP PREEMPT Debian 1:6.6.31-1+rpt1 (2024-05-29) aarch64 GNU/Linux
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by neolit123** [comment]
> @SlyMarbo, Please try to ask in the support channels for resolution as Github is not the right place to ask support queries. https://github.com/kubernetes/kubernetes/blob/master/SUPPORT.md

+1
seems like distro problem that we can't do much about.
you have the option to install "without a package manager"
https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/install-kubeadm/#installing-kubeadm-kubelet-and-kubectl

/close

**Comment by neolit123** [comment]
/reopen
actually given these package names let's get feedback from the deb/rpm experts.

> Depends: socat:arm64 but it is not installable

parhaps that's something that can be fixed on the pkg side.

**Comment by SlyMarbo** [comment]
I think I've diagnosed part of the cause. It turns out that the Raspbian installation I received has a 64-bit kernel but a 32-bit userspace. So presumably the package is using the wrong logic to determine which version of the package to install and is picking `arm64`, which then fails to install because there's no 64-bit `libc`, etc. I'm going to build a new Raspbian image that's entirely 64-bit.

**Comment by SlyMarbo** [comment]
Yes, that was my mistake; I discovered subsequently that I was installing 1.28 for some reason. Changing to 1.31 doesn't seem to have changed things. Separately, I've now reinstalled the Raspberry Pi with a fully 64-bit installation and that's working. So I think the issue remains that a 64-bit kernel with a 32-bit userspace results in an attempt to install the 64-bit packages, resulting in the error shown.

**Comment by xmudrii** [comment]
> So I think the issue remains that a 64-bit kernel with a 32-bit userspace results in an attempt to install the 64-bit packages, resulting in the error shown.

@kubernetes/release-engineering @neolit123 I think this sounds reasonable, i.e. we shouldn't support a 32-bit userspace on the 64-bit operating system. That said, I don't think we should change anything on our side, but looking for some feedback on that from y'all.

@aojea We indeed require `socat` in [our kubelet package](https://github.com/kubernetes/release/blob/a53bb39059bd6e51e58c3f0d058d22d06fae86e2/cmd/krel/templates/latest/kubelet/kubelet.spec#L31). Just to confirm, kubelet doesn't require it any longer and we should be able to remove it?

**Comment by saschagrunert** [comment]
The issue seems more general than #3722 as outlined by @SlyMarbo. I assume that other dependencies are still throwing an error after that PR got merged:

```
ethtool:arm64 but it is not installable
conntrack:arm64 but it is not installable
libc6:arm64 but it is not going to be installed
```

**Comment by xmudrii** [comment]
@saschagrunert That's correct, but my point was that I'm not sure if we should support this case at all. If I understand correctly, this is happening only if you use a 32-bit userspace. I don't see it being a good idea to mix and match 64-bit kernel and 32-bit packages unless I'm missing something. And we don't provide support for 32-bit at all.

**Comment by aojea** [comment]
> @aojea We indeed require `socat` in [our kubelet package](https://github.com/kubernetes/release/blob/a53bb39059bd6e51e58c3f0d058d22d06fae86e2/cmd/krel/templates/latest/kubelet/kubelet.spec#L31). Just to confirm, kubelet doesn't require it any longer and we should be able to remove it?

socat neither ebtables since the dockershim deprecation that I'm aware of, containerd and crio has a native socat implementation for port-forwarding and ebtables was a kubenet legacy thing IIRC
